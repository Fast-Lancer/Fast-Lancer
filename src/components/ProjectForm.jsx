import { useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { useUser } from '../context/UserContext.jsx'
import { getClients } from '../services/clients.js'
import { createProject, updateProject } from '../services/projects.js'

export default function ProjectForm() {

  const [startDate, setStartDate] = useState('')
  const [endDate, setEndDate] = useState('')
  const [hourlyRate, setHourlyRate] = useState('')
  const [hoursWorked, setHoursWorked] = useState('')
  const [URL, setURL] = useState('')
  const [description, setDescription] = useState('')
  const [completedAt, setCompletedAt] = useState('')
  const [priceQuoted, setPriceQuoted] = useState('')
  const [hoursQuoted, setHoursQuoted] = useState('')
  const [title, setTitle] = useState('')
  const [notes, setNotes] = useState('')
  const [clientId, setClientId] = useState('')
  const [clientsAvailable, setClientsAvailable] = useState([{ client_name: '', id: '' }])
  const { user } = useUser()
  const history = useHistory()

  useEffect(() => {
    getClients().then(setClientsAvailable)
  }, [])

  const handleSubmit = (e) => {
    e.preventDefault()

    createProject({ date_start: startDate, date_end: endDate, hourly_rate: hourlyRate, hours_worked: hoursWorked, client_id: Number(clientId), URL, description, completed_at: completedAt, price_quoted: priceQuoted, hours_quoted: hoursQuoted, title, notes, user_id: user.id }).then(project => history.push(`/projects/${project[0].id}`))
  }

  return <form onSubmit={handleSubmit}>
    <label htmlFor='startDate'>Start Date:</label>
    <input name='startDate' value={startDate} type='date' onChange={({ target }) => setStartDate(target.value)}/>

    <label htmlFor='endDate'>End Date:</label>
    <input name='endDate' value={endDate} type='date' onChange={({ target }) => setEndDate(target.value)}/>

    <label htmlFor='hourlyRate'>Hourly Rate:</label>
    <input name='hourlyRate' value={hourlyRate} type='text' onChange={({ target }) => setHourlyRate(target.value)}/>

    <label htmlFor='hoursWorked'>Hours Worked:</label>
    <input name='hoursWorked' value={hoursWorked} type='text' onChange={({ target }) => setHoursWorked(target.value)}/>

    <label htmlFor='URL'>URL:</label>
    <input name='URL' value={URL} type='text' onChange={({ target }) => setURL(target.value)}/>

    <label htmlFor='description'>Description:</label>
    <input name='description' value={description} type='text' onChange={({ target }) => setDescription(target.value)}/>

    <label htmlFor='completedAt'>Completed At:</label>
    <input name='completedAt' value={completedAt} type='date' onChange={({ target }) => setCompletedAt(target.value)}/>
    
    <label htmlFor='priceQuoted'>Price Quoted:</label>
    <input name='priceQuoted' value={priceQuoted} type='text' onChange={({ target }) => setPriceQuoted(target.value)}/>

    <label htmlFor='hoursQuoted'>Hours Quoted:</label>
    <input name='hoursQuoted' value={hoursQuoted} type='text' onChange={({ target }) => setHoursQuoted(target.value)}/>

    <label htmlFor='title'>Title:</label>
    <input name='title' value={title} type='text' onChange={({ target }) => setTitle(target.value)}/>

    <label htmlFor='notes'>Notes:</label>
    <input name='notes' value={notes} type='text' onChange={({ target }) => setNotes(target.value)}/>

    <label htmlFor='client'>Client: </label>
    <select name='client' value={clientId} onChange={({ target }) => setClientId(target.value)}>
      {clientsAvailable.map(client => <option key={client.id} value={client?.id}>{client?.client_name}</option>)}
    </select>

    <button type="submit">Submit</button>
  </form>
}
