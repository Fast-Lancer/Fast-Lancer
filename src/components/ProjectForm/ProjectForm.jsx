import {  useState } from 'react'
import { useUser } from '../../context/UserContext.jsx'

import changeValue from '../../utils/changeValue.js'

export default function ProjectForm({ handleProject, clientsAvailable, initialValues }) {
  const [form, setForm] = useState(initialValues)
  const { user } = useUser()
  
  const handleSubmit = (e) => {
    e.preventDefault()

    // replace empty strings with null in form object without mutation by creating a two dimensional array to iterate through, then make it back into object
    const nullifiedForm = changeValue(form, null)

    handleProject({ ...nullifiedForm, user_id: user.id })
  }

  const handleChange = (e) => {
    const { name, value } = e.target

    setForm(prev => ({ ...prev, [name]: value }))
  }

  return <form onSubmit={handleSubmit}>
    <label htmlFor='date_start'>Start Date:</label>
    <input name='date_start' value={form.date_start} type='date' onChange={handleChange}/>

    <label htmlFor='date_end'>End Date:</label>
    <input name='date_end' value={form.date_end} type='date' onChange={handleChange}/>

    <label htmlFor='hourly_rate'>Hourly Rate:</label>
    <input name='hourly_rate' value={form.hourly_rate} type='text' onChange={handleChange}/>

    <label htmlFor='hours_worked'>Hours Worked:</label>
    <input name='hours_worked' value={form.hours_worked} type='text' onChange={handleChange}/>

    <label htmlFor='URL'>URL:</label>
    <input name='URL' value={form.URL} type='text' onChange={handleChange}/>

    <label htmlFor='description'>Description:</label>
    <input name='description' value={form.description} type='text' onChange={handleChange}/>

    <label htmlFor='completed_at'>Completed At:</label>
    <input name='completed_at' value={form.completed_at} type='date' onChange={handleChange}/>
    
    <label htmlFor='price_quoted'>Price Quoted:</label>
    <input name='price_quoted' value={form.price_quoted} type='text' onChange={handleChange}/>

    <label htmlFor='hours_quoted'>Hours Quoted:</label>
    <input name='hours_quoted' value={form.hours_quoted} type='text' onChange={handleChange}/>

    <label htmlFor='title'>Title:</label>
    <input name='title' value={form.title} type='text' onChange={handleChange}/>

    <label htmlFor='notes'>Notes:</label>
    <input name='notes' value={form.notes} type='text' onChange={handleChange}/>

    <label htmlFor='client_id'>Client: </label>
    <select name='client_id' value={form.client_id} onChange={handleChange}>
      <option value={''}>required</option>
      {clientsAvailable.map(client => <option key={client.id} value={client?.id}>{client?.client_name}</option>)}
    </select>

    <button type="submit">Submit</button>
  </form>
}
