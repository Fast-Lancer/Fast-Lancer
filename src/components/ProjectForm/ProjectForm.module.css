.buttonWrapper {
  display: flex;
  justify-content: flex-start;
}

.buttonWrapper button {
  margin: 1vw;
}
