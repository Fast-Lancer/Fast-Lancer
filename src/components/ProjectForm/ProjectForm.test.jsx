import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import ProjectForm from './ProjectForm.jsx'


it('should display a form with empty field values', () => {
  const { container } = render(
    <MemoryRouter>
      <ProjectForm initialValues={{
        date_start: '', date_end: '', hourly_rate: '', hours_worked: '', URL: '', description: '', completed_at: '', price_quoted: '', hours_quoted: '', title: '', notes: '', client_id: '', user_id: ''
      }} clientsAvailable={[{ client_name: 'person1', id: '1' }, { client_name: 'person2', id: '2' }]} />
    </MemoryRouter>
  )

  expect(container).toMatchSnapshot()
})

it('should display a form with provided values', () => {
  const handleSubmit = jest.fn()

  const { container } = render(
    <MemoryRouter>
      <ProjectForm initialValues={{
        date_start: '2022-01-20', date_end: '2022-01-21', hourly_rate: '30', hours_worked: '50', URL: 'some@url.com', description: 'testing', completed_at: '2022-01-21', price_quoted: '4000', hours_quoted: '100', title: 'testing', notes: 'testing', client_id: '1', user_id: ''
      }} clientsAvailable={[{ client_name: 'person1', id: '1' }, { client_name: 'person2', id: '2' }]} handleProject={handleSubmit}/>
    </MemoryRouter>
  )

  expect(container).toMatchSnapshot()
})
