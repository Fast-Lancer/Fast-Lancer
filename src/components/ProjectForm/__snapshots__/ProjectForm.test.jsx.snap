// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a form with empty field values 1`] = `
<div>
  <form>
    <label
      for="date_start"
    >
      Start Date:
    </label>
    <input
      name="date_start"
      type="date"
      value=""
    />
    <label
      for="date_end"
    >
      End Date:
    </label>
    <input
      name="date_end"
      type="date"
      value=""
    />
    <label
      for="hourly_rate"
    >
      Hourly Rate:
    </label>
    <input
      name="hourly_rate"
      type="text"
      value=""
    />
    <label
      for="hours_worked"
    >
      Hours Worked:
    </label>
    <input
      name="hours_worked"
      type="text"
      value=""
    />
    <label
      for="URL"
    >
      URL:
    </label>
    <input
      name="URL"
      type="text"
      value=""
    />
    <label
      for="description"
    >
      Description:
    </label>
    <input
      name="description"
      type="text"
      value=""
    />
    <label
      for="completed_at"
    >
      Completed At:
    </label>
    <input
      name="completed_at"
      type="date"
      value=""
    />
    <label
      for="price_quoted"
    >
      Price Quoted:
    </label>
    <input
      name="price_quoted"
      type="text"
      value=""
    />
    <label
      for="hours_quoted"
    >
      Hours Quoted:
    </label>
    <input
      name="hours_quoted"
      type="text"
      value=""
    />
    <label
      for="title"
    >
      Title:
    </label>
    <input
      name="title"
      type="text"
      value=""
    />
    <label
      for="notes"
    >
      Notes:
    </label>
    <input
      name="notes"
      type="text"
      value=""
    />
    <label
      for="client_id"
    >
      Client: 
    </label>
    <select
      name="client_id"
    >
      <option
        value=""
      >
        required
      </option>
      <option
        value="1"
      >
        person1
      </option>
      <option
        value="2"
      >
        person2
      </option>
    </select>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should display a form with provided values 1`] = `
<div>
  <form>
    <label
      for="date_start"
    >
      Start Date:
    </label>
    <input
      name="date_start"
      type="date"
      value="2022-01-20"
    />
    <label
      for="date_end"
    >
      End Date:
    </label>
    <input
      name="date_end"
      type="date"
      value="2022-01-21"
    />
    <label
      for="hourly_rate"
    >
      Hourly Rate:
    </label>
    <input
      name="hourly_rate"
      type="text"
      value="30"
    />
    <label
      for="hours_worked"
    >
      Hours Worked:
    </label>
    <input
      name="hours_worked"
      type="text"
      value="50"
    />
    <label
      for="URL"
    >
      URL:
    </label>
    <input
      name="URL"
      type="text"
      value="some@url.com"
    />
    <label
      for="description"
    >
      Description:
    </label>
    <input
      name="description"
      type="text"
      value="testing"
    />
    <label
      for="completed_at"
    >
      Completed At:
    </label>
    <input
      name="completed_at"
      type="date"
      value="2022-01-21"
    />
    <label
      for="price_quoted"
    >
      Price Quoted:
    </label>
    <input
      name="price_quoted"
      type="text"
      value="4000"
    />
    <label
      for="hours_quoted"
    >
      Hours Quoted:
    </label>
    <input
      name="hours_quoted"
      type="text"
      value="100"
    />
    <label
      for="title"
    >
      Title:
    </label>
    <input
      name="title"
      type="text"
      value="testing"
    />
    <label
      for="notes"
    >
      Notes:
    </label>
    <input
      name="notes"
      type="text"
      value="testing"
    />
    <label
      for="client_id"
    >
      Client: 
    </label>
    <select
      name="client_id"
    >
      <option
        value=""
      >
        required
      </option>
      <option
        value="1"
      >
        person1
      </option>
      <option
        value="2"
      >
        person2
      </option>
    </select>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
